<a href="https://colab.research.google.com/github/ZaKSa/neural-networks-8semester-/blob/marathon/2_%D0%92%D0%B2%D0%B5%D0%B4%D0%B5%D0%BD%D0%B8%D0%B5_%D0%B2_%D0%BE%D0%B1%D1%83%D1%87%D0%B5%D0%BD%D0%B8%D0%B5_%D0%BD%D0%B5%D0%B9%D1%80%D0%BE%D0%BD%D0%BD%D1%8B%D1%85_%D1%81%D0%B5%D1%82%D0%B5%D0%B9.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


Функция (Function) потерь может дать бо́льшую практическую гибкость вашим Нейронным сетям (Neural Network) и будет определять, как именно выходные данные связаны с исходными.

Среднеквадратичная ошибка (MSE) — одна из основных функций расчёта отклонения. Для каждой точки вычисляется квадрат отклонения, после чего полученные значения суммируются и делятся на общее количество точек. Чем ближе полученное значение к нулю, тем точнее наша модель. Данный метод расчёта в значительной мере чувствителен к выбросам в выборке, или к выборкам где разброс значений очень большой. В основном, данная функция применяется для переменных, распределение которых близко к распределению Гаусса.

Средняя абсолютная ошибка (MAE) – это усреднённая сумма модулей разницы между реальным и предсказанным значениями. MAE во многом похожа на MSE, но она отличается меньшей чувствительностью к выбросам значений (так как не берётся квадрат отклонения).

Кросс-энтропия (Перекрестная энтропия) – это Функция потерь (Loss Function), которую можно использовать для количественной оценки разницы между двумя Распределениями вероятностей (Probability Distribution).

Среднеквадратичная логарифмическая ошибка (MSLE) – усреднённая сумма квадратов разностей между логарифмами значений. Благодаря большому гасящему эффекту логарифма она более применима к моделям, строящимся на данных, которые имеют большой разброс значений на несколько порядков.

1. Расчет среднеквадратической ошибки





In [ ]:
weight = 0.5
input = 0.5
goal_prediction = 0.8

pred = input * weight
error = (pred - goal_prediction) ** 2
print(error)

0.30250000000000005


2. Попробуем обучить сеть, изменяя вес в разные стороны и рассчитывая значение ошибки для каждого изменения

In [ ]:
weight = 0.5
input = 0.5
goal_prediction = 0.8

step_amount = 0.0001


for iteration in range(1101):

    prediction = input * weight
    error = (prediction - goal_prediction) ** 2

    print("Error:" + str(error) + " Prediction:" + str(prediction))
    
    up_prediction = input * (weight + step_amount) #попробовать увеличить
    up_error = (goal_prediction - up_prediction) ** 2

    down_prediction = input * (weight - step_amount) #попробовать уменьшить
    down_error = (goal_prediction - down_prediction) ** 2

    if(down_error < up_error):
        weight = weight - step_amount #если уменьшение дало лучший результат, уменьшить
        
    if(down_error > up_error):
        weight = weight + step_amount # если увеличение дало лучший результат, увеличить

Error:0.30250000000000005 Prediction:0.25
Error:0.30244500250000006 Prediction:0.25005
Error:0.30239001000000004 Prediction:0.2501
Error:0.30233502250000005 Prediction:0.25015
Error:0.3022800400000001 Prediction:0.2502
Error:0.3022250625000001 Prediction:0.25025
Error:0.30217009000000006 Prediction:0.25029999999999997
Error:0.3021151225000001 Prediction:0.25034999999999996
Error:0.3020601600000001 Prediction:0.25039999999999996
Error:0.3020052025000001 Prediction:0.25044999999999995
Error:0.3019502500000001 Prediction:0.25049999999999994
Error:0.30189530250000013 Prediction:0.25054999999999994
Error:0.30184036000000014 Prediction:0.25059999999999993
Error:0.3017854225000001 Prediction:0.25064999999999993
Error:0.30173049000000013 Prediction:0.2506999999999999
Error:0.30167556250000016 Prediction:0.2507499999999999
Error:0.30162064000000016 Prediction:0.2507999999999999
Error:0.30156572250000013 Prediction:0.2508499999999999
Error:0.3015108100000002 Prediction:0.2508999999999999
Error:0

**Вывод об изменении количества итераций**

Первоначально было 1101 итераций. две последние ошибки: *Error:2.5000000003280753e-07 Prediction:0.7994999999999672
Error:1.0799505792475652e-27 Prediction:0.7999999999999672*

Уменьшим в два раза количество итераций: *Error:0.07617600000000137 Prediction:0.5239999999999976
Error:0.07590025000000139 Prediction:0.5244999999999975*

Увеличим в два раза:*Error:1.0799505792475652e-27 Prediction:0.7999999999999672
Error:2.4999999996708233e-07 Prediction:0.8004999999999671*

Если уменьшить количество итераций, то точность уменьшается. Если увеличить количество, то предсказание в пределах машинного нуля не изменяется 



Неэффективность данного кода: вычисляем прогноз три раза

Сделаем теперь именим влеичину шага на 0,2, получим: *Error:0.0025000000000000157 Prediction:0.7499999999999999
Error:0.0024999999999999823 Prediction:0.8499999999999999*. 

Потом на 10: *Error:19.802500000000002 Prediction:5.25
Error:0.30250000000000005 Prediction:0.25* - предсказание не меняется. 

Уменьшим шаг до  0.0001:*Error:0.24507450250000604 Prediction:0.30494999999999395
Error:0.24502500000000604 Prediction:0.30499999999999394*

> В данном примере сеть достигнет идеального значения веса,
если оно равно n*step_amount. Иначе при выбранном
значении шага сеть будет колебаться вокруг целевого
значения в меньшую или большую сторону, разница между
прогнозом и эталонным значением будет меньше величины
шага.

Используем производную функцию потерь. В многомерном случае - это вектор градиента, состоящих из частных производных по каждому из весов сети. Градиент - направление, в котором функция быстрее всего возрастает

наклон смотрим

если знак производной -, увеличим вес, при + - уменьшаем. Меняем вес в направлении антиградиента на величину произвожной в текущей точке.

3. Реализуйте градиентный спуск: определите направление и величину изменения веса, используя производную функции потерь для текущего веса . Выполните 20 итераций градиентного спуска . 
 weight = weight - derivative

вычислим наклон в точке.
если он отрицательный, двигаемся вправо.
если он положительный, двигаемся влево.
повторим, пока он не будет равен 0

In [ ]:
weight = 0.5
goal_prediction = 0.8
input = 0.5

for iteration in range(20):
  loss=(input*weight-goal_prediction)**2 #функция потерь
  derivative=2*input*(input*weight-goal_prediction) #производная от ф потерь по весам
  print(derivative,loss) #проверочка
  if derivative<0:
    weight = weight - derivative
  else:
    weight = weight + derivative
print(weight)



-0.55 0.30250000000000005
-0.275 0.07562500000000001
-0.13749999999999996 0.01890624999999999
-0.06874999999999998 0.004726562499999997
-0.034375000000000044 0.001181640625000003
-0.017187500000000022 0.00029541015625000075
-0.008593749999999956 7.385253906249924e-05
-0.004296874999999978 1.846313476562481e-05
-0.0021484375000000444 4.6157836914064404e-06
-0.0010742187500000222 1.1539459228516101e-06
-0.0005371093749999556 2.884864807128429e-07
-0.0002685546874999778 7.212162017821073e-08
-0.0001342773437500444 1.803040504456759e-08
-6.71386718750222e-05 4.507601261141898e-09
-3.356933593745559e-05 1.1269003152817474e-09
-1.6784667968727796e-05 2.8172507882043685e-10
-8.392333984419409e-06 7.043126970604095e-11
-4.1961669922097045e-06 1.760781742651024e-11
-2.098083496049341e-06 4.401954356394625e-12
-1.0490417480246705e-06 1.1004885890986563e-12
1.599998950958252


-0.55 0.30250000000000005

-0.275 0.07562500000000001

-0.13749999999999996 0.01890624999999999

-0.06874999999999998 0.004726562499999997

-0.034375000000000044 0.001181640625000003

-0.017187500000000022 0.00029541015625000075

-0.008593749999999956 7.385253906249924e-05

-0.004296874999999978 1.846313476562481e-05

-0.0021484375000000444 4.6157836914064404e-06

-0.0010742187500000222 1.1539459228516101e-06

-0.0005371093749999556 2.884864807128429e-07

-0.0002685546874999778 7.212162017821073e-08

-0.0001342773437500444 1.803040504456759e-08

-6.71386718750222e-05 4.507601261141898e-09

-3.356933593745559e-05 1.1269003152817474e-09

-1.6784667968727796e-05 2.8172507882043685e-10

-8.392333984419409e-06 7.043126970604095e-11

-4.1961669922097045e-06 1.760781742651024e-11

-2.098083496049341e-06 4.401954356394625e-12

-1.0490417480246705e-06 1.1004885890986563e-12

1.599998950958252

СДЕЛАНО


и совпало со значениями из программной реализации



шарик катится в самую глубокую яму - в направлении антиградиента к поверхности

In [ ]:
## input=2, goal=0,8

weight = 0.5
goal_prediction = 0.8
input = 2

for iteration in range(20):
  loss=(input*weight-goal_prediction)**2 #функция потерь
  derivative=2*input*(input*weight-goal_prediction) #производная от ф потерь по весам
  print(derivative,loss) #проверочка
  if derivative<0:
    weight = weight - derivative
  else:
    weight = weight + derivative
print(weight)



0.7999999999999998 0.03999999999999998
7.199999999999998 3.2399999999999984
64.79999999999998 262.4399999999999
583.1999999999998 21257.63999999999
5248.799999999998 1721868.839999999
47239.19999999999 139471376.03999993
425152.7999999999 11297181459.239994
3826375.199999999 915071698198.4395
34437376.79999999 74120807554073.6
309936391.1999999 6003785411879962.0
2789427520.7999997 4.8630661836227706e+17
25104847687.199997 3.939083608734444e+19
225943629184.79996 3.190657723074899e+21
2033492662663.1997 2.5844327556906685e+23
18301433963968.797 2.0933905321094415e+25
164712905675719.2 1.695646331008648e+27
1482416151081472.8 1.3734735281170048e+29
1.3341745359733254e+16 1.1125135577747738e+31
1.200757082375993e+17 9.01135981797567e+32
1.0806813741383936e+18 7.299201452560292e+34
1.2157665459056927e+18


Получился очень большой целевой вес

In [ ]:
#изменим weight
# goal_pr
#input

weight = 0.5#уменьшим и увеличим в 10 раз
goal_prediction = 8#уменьшим и увеличим в 10 раз
input = 0.5#уменьшим и увеличим в 10 раз


for iteration in range(20):
  loss=(input*weight-goal_prediction)**2 #функция потерь
  derivative=2*input*(input*weight-goal_prediction) #производная от ф потерь по весам
  print(derivative,loss) #проверочка
  if derivative<0:
    weight = weight - derivative
  else:
    weight = weight + derivative
print(weight)

#целевой вес до изменений 1.599998950958252
#input = 0.05: [уменьшим в 10 раз] 1.978537555743427
#input = 5: 4.818173548056033e+33
#weight = 0.05: 1.5999985218048096
#weight = 5: 11307.472882270813
#goal_prediction = 0.08: 1130.7472882270806
#goal_prediction = 8: 15.999985218048096

-7.75 60.0625
-3.875 15.015625
-1.9375 3.75390625
-0.96875 0.9384765625
-0.484375 0.234619140625
-0.2421875 0.05865478515625
-0.12109375 0.0146636962890625
-0.060546875 0.003665924072265625
-0.0302734375 0.0009164810180664062
-0.01513671875 0.00022912025451660156
-0.007568359375 5.728006362915039e-05
-0.0037841796875 1.4320015907287598e-05
-0.00189208984375 3.5800039768218994e-06
-0.000946044921875 8.950009942054749e-07
-0.0004730224609375 2.2375024855136871e-07
-0.00023651123046875 5.593756213784218e-08
-0.000118255615234375 1.3984390534460545e-08
-5.91278076171875e-05 3.496097633615136e-09
-2.956390380859375e-05 8.74024408403784e-10
-1.4781951904296875e-05 2.18506102100946e-10
15.999985218048096


проблема расхождения

если входное значение слишком велико происходит значительное увеличение веса даже при маленькой ошибке. избыточная коререкция веса сети


при больших входных данных прогноз сильно чувствителен к изменению веса, небольшое изменение веса вызывает значительное изменение в прогнозе

**гиперпараметр - скорость обучения**

Значение скорости обучения определим экспериментально: смотрим на изменение ошибки. 

Если наблюдается расхождение, то уменьшим параметр. 

Если обучение происходит медленно, то увеличим скорость 

4. Реализуйте алгоритм градиентного спуска, добавив параметр скорости обучения.

In [ ]:
weight = 0.5
goal_prediction = 0.8
input = 0.5
learning_rate = 0.00001

for iteration in range(20):
  loss=(input*weight-goal_prediction)**2 #функция потерь
  derivative=2*input*(input*weight-goal_prediction) #производная от ф потерь по весам
  print(derivative,loss) #проверочка
  if derivative<0:
    weight = weight - (derivative)*learning_rate
  else:
    weight = weight + (derivative)*learning_rate
print(weight)

#learning_rate = 10: [вес] 2681182856046184.0
#learning_rate = 1:[вес] 1.599998950958252
#learning_rate = 0.1: [вес] 1.2056654853506037
#learning_rate = 0.01:[вес] 0.6049284716979204
#learning_rate = 0.001: [вес] 0.5109479064174384
#learning_rate = 0.0001: [вес] 0.5010994776567166
#Если праметр <=0.01, то ошибка становится больше чем 0,3 и почти не изменяется потом
#при 0.5 ошибка 0.284444444444509

-0.55 0.30250000000000005
-0.5499972500000001 0.3024969750075626
-0.54999450001375 0.3024939500453749
-0.54999175004125 0.30249092511343684
-0.5499890000824998 0.30248790021174804
-0.5499862501374994 0.302484875340308
-0.5499835002062486 0.3024818504991167
-0.5499807502887476 0.30247882568817375
-0.5499780003849961 0.3024758009074788
-0.5499752504949942 0.3024727761570316
-0.5499725006187417 0.30246975143683186
-0.5499697507562387 0.3024667267468793
-0.5499670009074848 0.30246370208717344
-0.5499642510724804 0.3024606774577142
-0.549961501251225 0.3024576528585012
-0.5499587514437188 0.302454628289534
-0.5499560016499616 0.3024516037508126
-0.5499532518699533 0.30244857924233626
-0.5499505021036939 0.302445554764105
-0.5499477523511834 0.30244253031611856
0.5001099947751567


In [ ]:
weight = 0.5
goal_prediction = 0.8
input = 0.5
learning_rate = 0.01

for iteration in range(20):
  loss=(input*weight-goal_prediction)**2 #функция потерь
  derivative=2*input*(input*weight-goal_prediction) #производная от ф потерь по весам
  print(derivative,loss) #проверочка
  if derivative<0:
    weight = weight - (derivative)*learning_rate
  else:
    weight = weight + (derivative)*learning_rate
print(weight)

-0.55 0.30250000000000005
-0.54725 0.2994825625
-0.5445137500000001 0.29649522393906264
-0.54179118125 0.29353768408027037
-0.5390822253437502 0.29060964568156983
-0.5363868142170314 0.28771081446589614
-0.5337048801459463 0.28484089909159893
-0.5310363557452165 0.2819996111231601
-0.5283811739664904 0.2791866650022066
-0.5257392680966579 0.27640177801880955
-0.5231105717561747 0.273644670283072
-0.5204950188973938 0.27091506469699833
-0.5178925438029068 0.2682126869266458
-0.5153030810838923 0.2655372653745525
-0.5127265656784727 0.2628885311524412
-0.5101629328500805 0.26026621805419575
-0.5076121181858301 0.2576700625291051
-0.5050740575949009 0.25509980365537727
-0.5025486873069265 0.25255518311391495
-0.5000359438703919 0.25003594516235367
0.6049284716979204


Что происходит при input=2?

In [ ]:
weight = 0.5
goal_prediction = 0.8
input = 2
learning_rate = 0.1 #минимальная ошибка

for iteration in range(20):
  loss=(input*weight-goal_prediction)**2 #функция потерь
  derivative=2*input*(input*weight-goal_prediction) #производная от ф потерь по весам
  print(derivative,loss) #проверочка
  if derivative<0:
    weight = weight - (derivative)*learning_rate
  else:
    weight = weight + (derivative)*learning_rate
print(weight)

0.7999999999999998 0.03999999999999998
1.4399999999999995 0.1295999999999999
2.5919999999999996 0.4199039999999999
4.6655999999999995 1.3604889599999996
8.39808 4.4079842304
15.116544000000001 14.281868906496003
27.209779200000003 46.273255257047055
48.97760256000001 149.92534703283246
88.15968460800002 485.75812438637723
158.68743229440005 1573.8563230118625
285.63737812992014 5099.294486558436
514.1472806338562 16521.714136449333
925.4651051409412 53530.35380209584
1665.837189253694 173438.34631879049
2998.50694065665 561940.2420728814
5397.31249318197 1820686.384316136
9715.162487727546 5899023.885184281
17487.292477909585 19112837.387997072
31477.126460237254 61925593.13711052
56658.82762842706 200638921.76423812
12748.636216396088


Ситуация стала намного лучше, оптимальный вес 0.18830167591902072

ошибка 0.1795378537174714

Параметр скорректировал полученные данные

In [ ]:
weight = 0.5
goal_prediction = 0.8
input = 2
learning_rate = 0.01

for iteration in range(20):
  loss=(input*weight-goal_prediction)**2 #функция потерь
  derivative=2*input*(input*weight-goal_prediction) #производная от ф потерь по весам
  print(derivative,loss) #проверочка
  if derivative<0:
    weight = weight - (derivative)*learning_rate
  else:
    weight = weight + (derivative)*learning_rate
print(weight)

0.7999999999999998 0.03999999999999998
0.8639999999999999 0.04665599999999999
0.9331199999999997 0.05441955839999997
1.0077695999999996 0.06347497291775996
1.0883911679999994 0.07403720841127517
1.1754624614399996 0.0863569998909114
1.2694994583551997 0.10072680467275909
1.3710594150236153 0.11748774497030613
1.4807441682255043 0.13703770573336502
1.5992037016835443 0.1598407799673969
1.7271399978182282 0.18643828575397184
1.8653111976436865 0.21746161650343274
2.014536093455181 0.2536472294896039
2.175698980931595 0.29585412847667386
2.349754899406123 0.3450842554551924
2.5377352913586124 0.40250627556293633
2.7407541146673013 0.4694833198166089
2.9600144438406852 0.5476053442340926
3.19681559934794 0.6387268735146456
3.452560847295775 0.7450110252674824
0.8660957143849296
